p4-fusion/commands/file_data.cc

```diff
@@ -53,24 +59,43 @@ void FileData::SetFromDepotFile(const std::string& fromDepotFile, const std::str
 	}
 }
 
-void FileData::MoveContentsOnceFrom(const std::vector<char>& contents)
+void FileData::StartWrite()
 {
-	// TODO double-check the thread logic here.  It needs to be thread safe.
+	MTR_SCOPE("FileData", __func__);
 
-	if (m_data->isContentsSet)
+	if (m_data->isBlobOIDSet)
 	{
 		// Do not set the contents.  Assume that
 		// they were already set or, worst case, are currently being set.
 		return;
 	}
-	m_data->isContentsSet = true;
-	m_data->contents = std::move(contents);
+
+	GIT2(git_repository_open_bare(&repo, FileData::repoPath.c_str()));
+	GIT2(git_blob_create_from_stream(&writer, repo, nullptr));
+}
+
+void FileData::Write(const char* contents, int length)
+{
+	MTR_SCOPE("FileData", __func__);
+
+	GIT2(writer->write(writer, contents, length));
+}
```

Comment:
This piece writes small-ish chunks of data to the blob one-by-one, so memory consumption is very limited. Writing a large 1GB binary did not cause any noticeable change in the memory consumption of the process.